### ISSUE TITLE
Improve ergonomics of importing multiple items from a single module

### ISSUE THREAD

**Issue by colinwahl** [natefaubion/purescript-tidy-codegen#3]
Earlier today we discussed possibly improving the ergonomics of importing multiple items from a single module.

Right now, you have to have multiple `importFrom` calls, like:

```
lookupIdent <- importFrom "Data.Map" (importValue "lookup")
insertIdent <- importFrom "Data.Map" (importValue "insert")
```

which then gets normalized in the generated code to be

```
import Data.Map (lookup, insert)
```

It would be nice if we could use records to import multiple values more similarly to how we do in source code, something like:

```
dataMap@{ insert, lookup } <- importFrom "Data.Map"
  { lookup: importValue "lookup"
  , insert: importValue "insert"
  }
```

where `dataMap.insert` and `dataMap.lookup` then correspond to the identifiers for `insert` and `lookup` imported from `Data.Map`.

**Comment by natefaubion** [comment]
Note that this is equivalent to `traverseRecord` (if that exists somewhere).

**Comment by JordanMartinez** [comment]
To enable `traverseRecord`, we'd need to define `functorRecord`, `foldableRecord`, and then `traverseRecord`. Because the kind of a record is `Row Type -> Type` rather than `Type -> Type`, we can't define a Functor instance for `Record`, which means we can't define a `Foldable` or `Traversable` instance.

However, Record instances typically call a function from some helper type class (e.g. `MonoidRecord`). So, we could instead define those helper type classes and then define a derived function that provides all the type class constraints needed to make that function work. For example,
```purescript
-- | Same as `Functor`'s `map` but works on homogenous records where
-- | each label is associated with the `Functor` type that stores the same `a` type:
-- |
-- | - Invalid, because the `Functor` type changes: `{ x :: Maybe Int, y :: Array Int }`
-- | - Invalid, because the `a` type changes: `{ x :: Array Int, y :: Array String }`
-- | - Valid: `{ a :: Array Int, b :: Array Int }`
-- |
-- | ```purescript
-- | mapRecord (_ + 1) { x: [ 1 ], y: [ 2] } == { x: [ 2 ], y: [ 3 ] }
-- | ```
mapRecord
  :: forall a b rowList rows mappedRows
   . RL.RowToList rows rowList
  => FunctorRecord rowList a b rows mappedRows
  => (a -> b)
  -> { | rows }
  -> { | mappedRows }
mapRecord func rec = mapRecordImpl (Proxy :: Proxy rowList) func rec

class FunctorRecord :: (RL.RowList Type) -> Type -> Type -> Row Type -> Row Type -> Constraint
class FunctorRecord rowlist a b row subrow | rowlist a b -> row subrow where
  mapRecordImpl :: Proxy rowlist -> (a -> b) -> { | row } -> { | subrow }

instance functorRecordNil :: FunctorRecord RL.Nil a b row () where
  mapRecordImpl _ _ _ = {}

instance functorRecordCons ::
  ( IsSymbol key
  , Row.Cons key (f b) subrowTail subrow
  , FunctorRecord rowlistTail a b row subrowTail
  , Functor f
  ) => FunctorRecord (RL.Cons key (f a) rowlistTail) a b row subrow where
  mapRecordImpl _ func rec = insert (map func (get rec)) tail
    where
    key = reflectSymbol (Proxy :: Proxy key)
    get = unsafeGet key
    insert = unsafeSet key :: f b -> { | subrowTail } -> { | subrow }
    tail = mapRecordImpl (Proxy :: Proxy rowlistTail) func rec
```

**Comment by natefaubion** [comment]
You can get `traverseRecord` using `record-extra`.

```
traverseRecord f =
  sequenceRecord <<< mapRecord f
```

I'm not suggesting that we add anything to Prelude. I'm also not sure that implementation is what people expect when they say they want to map over a record. They usually want to treat a homogeneous record as a container itself.

**Comment by JordanMartinez** [comment]
:facepalm: I thought I remembered seeing `sequenceRecord` before but couldn't remember where. Yeah, adding that to `record-extra` sounds more reasonable.

**Comment by natefaubion** [comment]
Fixed in 1.1.0, you can pass in a record to `importFrom`.
